### PR TITLE
lib: Move `atomic.h` arhitecture specific code to new `ukatomic`…

### PR DESCRIPTION
**Pull request by Starnox**
As part of the effort towards a platform re-arch, created a new design for an internal `uk` library that contains all the `arch` specific logic in different subdirectories and includes the right one using the build system.
Removed the `arch` prefix in all relevant locations.

### Prerequisite checklist

 - [X] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [X] Tested your changes against relevant architectures and platforms;
 - [X] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]

### Additional configuration

Select `ukatomic` library in `KConfig` menu.

### Description of changes

As part of the effort towards a platform re-arch, created a new design
for an internal `unikraft` library that contains all the `arch` specific
logic in different subdirectories and includes the right one using
the build system.

### CODE REVIEW

**Review by michpappas — CHANGES_REQUESTED** [review]
Hi @Starnox, please break down this PR into separate commits:
* One commit with that migration of atomic into own library.
* Separate commits for each updated calling library.

In general it is a good practice to update libraries in separate commits, as different libraries may have different maintainers. It also simplifies the review process in general.

Similarly for `plat`. Given that right now it's not so structured, you can make logic separations like `plat/common/lcpu`, `drivers/virtio`, and `plat/xen`.

Ditto for headers in `include/uk/`.

**Comment by Starnox** [comment]
> Hi @Starnox, please break down this PR into separate commits:
> 
>     * One commit with that migration of atomic into own library.
> 
>     * Separate commits for each updated calling library.
> 
> 
> In general it is a good practice to update libraries in separate commits, as different libraries may have different maintainers. It also simplifies the review process in general.
> 
> Similarly for `plat`. Given that right now it's not so structured, you can make logic separations like `plat/common/lcpu`, `drivers/virtio`, and `plat/xen`.
> 
> Ditto for headers in `include/uk/`.

Done! Let me know if the separation is satisfactory.

File: lib/ukatomic/Config.uk
Comment (michpappas) [inline lib/ukatomic/Config.uk:3]: A `help` section with some additional info on the contents of this library would be useful.

**Review by michpappas — CHANGES_REQUESTED** [review]
@Starnox thanks for the update, it looks good now 👍🏼 Some last nitpicks:

```
arch/lib: Move atomic operations logic from `arch` to new `ukatomic` library
```

Normally this denotes a hierarchy but there's no `lib` under `arch`. I think it's okay to just use `lib` here.

```    
    As part of the plat re-arch efforts, atomic operations will be implemented
    in an internal `uk` library.
```

I wouldn't mention the plat re-arch effort here, given that we there will be many more commits to come and certainly it would be too much to provide that context in all of them. A simple description of the changes (ie move headers into own library) would suffice.

In the rest of the commits:
```
    - Modify include path of `atomic.h`
    - Removed `arch prefix`
```
Since you use assertive voice in the commit title and the first bullet, you might as well use `Remove` instead of `Removed`.

Other than the above it looks great. Thanks again!

File: lib/ukatomic/Config.uk
Comment (michpappas) [inline lib/ukatomic/Config.uk]: This message can be improved a bit, as currently it provides either redundant information, or information not useful to the user. I explain:

> implemented for all supported architectures

Architecture support should be addressed by Kconfig dependencies, ie libraries not supported by a given architecture normally don't appear to the menu.

> This library is required by other libraries and applications, and should be enabled by default.

This is redundant as you already enable this by setting `default y`.

It's better to provide something as simple as `Atomic operations` or a more elaborate description like `Architecture-specific implementation of atomic operations. The set of operations is the same as the one provided by GCC's memory model aware atomic built-ins.`

**Review by michpappas — CHANGES_REQUESTED** [review]
Hi @Starnox sorry that I'll be ultra-pedantic here but I see that this PR is used as base for others like #959, so we better get this right.

In the commit message:
```
 lib: Move atomic operations logic from arch to new ukatomic library

- Move headers from `arch` directory to `ukatomic`
- Move header from `include` directory to `ukatomic`
- Provide config option which is by default selected
``` 

These bullets essentially provide the same information as `git diff`. A more concentrated description is normally preferred:
```
lib: Move atomic operations from arch to new ukatomic library

Introduce `lib/ukatomic` and migrate `ukatomic` implementation from `arch`.
This library is enabled by default.
```

or...

```
lib: Move atomic operations from arch to new ukatomic library

Migrate `ukatomic` implementation from `arch` into newly introduced `lib/ukatomic`.
This library is enabled by default.
```

...or some similar wording of your choice :)

**Comment by razvand** [comment]
@Starnox , could you please rebase this?

File: lib/ukatomic/Makefile.uk
Comment (michpappas) [inline lib/ukatomic/Makefile.uk]: Normally we don't put copyright in makefiles (at least in recently introduced libraries). You can have a look at recently introduced libraries like `uktimeconv`, `uknofault` or `ukvmem` for reference.

File: lib/ukatomic/arch/arm/include/uk/asm/atomic.h
Comment (michpappas) [inline lib/ukatomic/arch/arm/include/uk/asm/atomic.h]: Actually these are bit ops, not atomic, so they should not had been part of `atomic.h` in the first place. Perhaps you should move these into their own `ukbitops` library and leave `ukatomic` with what is in `include/uk/atomic.h`.

**Review by michpappas — CHANGES_REQUESTED** [review]
Hi @Starnox, thanks for updating this. Besides the comments below, please squash these commits together. Also apologies for requesting that because I know we previously asked to have these in separate commits, but there have been discussions during re-arch among maintainers where people expressed the opinion that for re-arch it makes more sense to have these changes atomically (for changes not related to re-arch the rule of one commit per library remains). Thanks a lot! 🥂

File: lib/ukatomic/arch/arm/include/uk/asm/atomic.h
Comment (Starnox) [reply lib/ukatomic/arch/arm/include/uk/asm/atomic.h]: I think this should have it's own separate PR after this is merged

**Comment by Starnox** [comment]
> Hi @Starnox, thanks for updating this. Besides the comments below, please squash these commits together. Also apologies for requesting that because I know we previously asked to have these in separate commits, but there have been discussions during re-arch among maintainers where people expressed the opinion that for re-arch it makes more sense to have these changes atomically (for changes not related to re-arch the rule of one commit per library remains). Thanks a lot! 🥂

Hi! I rebased, updated and tested this PR with the latest changes from staging

File: lib/ukatomic/exportsyms.uk
Comment (michpappas) [inline lib/ukatomic/exportsyms.uk]: These are now part of `ukbitops`, so they can be removed from `ukatomic`.

File: lib/ukatomic/Makefile.uk
Comment (michpappas) [inline lib/ukatomic/Makefile.uk]: Since `atomic.h` does not contain any directives that can be interpreted by the assembler, `ASINCLUDES` is strictly speaking not necessary.

File: lib/ukatomic/Config.uk
Comment (michpappas) [inline lib/ukatomic/Config.uk]: The linux kernel uses two spaces to indent `help`, and personally I prefer that too, but the Unikraft convention is to use a tab here too, so let's stick to that 🙂

File: lib/ukatomic/include/uk/atomic.h
Comment (michpappas) [inline lib/ukatomic/include/uk/atomic.h:36]: I'm leaving this here as github doesn't allow to comment on the closing brace: I think there is no reason to have `ACCESS_ONCE` and `UK_READ_ONCE` / `UK_WRITE_ONCE` outside the C++ guards. I checked on the commit message that introduced these changes and didn't see any specific reason mentioned, so I speculate it was an omission.

File: lib/ukbitops/Makefile.uk
Comment (michpappas) [inline lib/ukbitops/Makefile.uk]: Similarly to `atomic.h`, I doubt anything defined in `bitops.h` can be used by asm code, so all the `ASINCLUDES` in this file can go away.

File: lib/ukbitops/Makefile.uk
Comment (michpappas) [inline lib/ukbitops/Makefile.uk:4]: Since you are defining `CXXINCLUDES` here, you should be adding C++ guards in the headers.

File: lib/ukbitops/Config.uk
Comment (michpappas) [inline lib/ukbitops/Config.uk]: Same as `libukatomic`, please indent with a tab here.

File: lib/ukbitops/Makefile.uk
Comment (michpappas) [inline lib/ukbitops/Makefile.uk]: You can reduce these to a couple of lines by declaring:
```
CINCLUDES-$(CONFIG_LIBUKBITOPS)	 		+= -I$(LIBUKBITOPS_BASE)/arch/$(CONFIG_UK_ARCH)/include
CXXINCLUDES-$(CONFIG_LIBUKBITOPS)		+= -I$(LIBUKBITOPS_BASE)/arch/$(CONFIG_UK_ARCH)/include
```

File: lib/ukatomic/include/uk/atomic.h
Comment (michpappas) [inline lib/ukatomic/include/uk/atomic.h:71]: There is one instance of `ukarch_inc` remaining in `unixsock.c`:
```
lib/posix-unixsocket/unixsock.c:		if (!ukarch_inc(&data->listen.count))
```
Perhaps it was added after you did this work.

**Review by michpappas — CHANGES_REQUESTED** [review]
Hi @Starnox thank you very much for making these changes. A few comments below, I think once these are resolved we can merge.

**Review by michpappas — APPROVED** [review]
@Starnox thanks a lot for this work.

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request merged**
